synapse/replication/tcp/streams/events.py

```diff
@@ -176,16 +176,30 @@ async def _update_function(
             from_token, upper_limit, target_row_count
         )  # type: List[Tuple]
 
-        # again, if we've hit the limit there, we'll need to limit the other sources
-        assert len(state_rows) < target_row_count
+        assert len(state_rows) <= target_row_count
+
+        # there can be more than one row per stream_id in that table, so if we hit
+        # the limit there, we'll need to truncate the results so that we have a complete
+        # set of changes for all the stream IDs we include.
         if len(state_rows) == target_row_count:
             assert state_rows[-1][0] <= upper_limit
-            upper_limit = state_rows[-1][0]
-            limited = True
+            upper_limit = state_rows[-1][0] - 1
+
+            # search for the point to truncate the list
+            for idx in range(len(state_rows) - 1, 0, -1):
```

Comment:
I'd probably write this as `for idx, row in reversed(enumerate(state_rows)):` or something

Reply:
you can't do `reversed(enumerate(...))`, because enumerate returns a generator.

Reply:
Well that's tedious isn't it. Might be worth saying that we're walking backwards to find the last row of the set of rows with the second to last stream ID. Or something. But with better words. Mebs. :man_shrugging: